### PR TITLE
Module Serialization - layer-wise serialization support

### PR DESCRIPTION
**Pull request by wzhongyuan**
## What changes were proposed in this pull request?

Following the default serialization framework, this PR contains
1) Layer-wise serialization support
For most layers, just one declaration needs , e.g,  object Linear extends ModuleSerialization {...}
15 Layers override the serialization traits as they need to store  their own states, they are

BinaryTreeLSTM
BiRecurrent
Graph
MapTable
MaskedSelect
Recurrent
Reshape
Scale
SpatialContrastiveNormalization
SpatialDivisiveNormalization
SpatialFullConvolution
SpatialMaxPooling
SpatialSubtractiveNormalization
Transpose
VolumetricMaxPooling

2) Unit test cases for all layers 
3) another 2 unit tests added to DataConverter
4) Module/AbstractModule to use ModuleLoader/ModulePersister instead of java based serialization

## How was this patch tested?
Unit tests and common models tested locally

## Related links or issues (optional)
https://github.com/intel-analytics/BigDL/issues/608

### CODE REVIEW

**Comment by wzhongyuan** [comment]
@yiheng  this PR contains layer-wise support, but most modifications are very simple (just make the object extend from serialization trait) so you could go through very quickly.  Only 15 layers override the traits to store their states.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BinaryTreeLSTM.scala [hunk 1]
```diff
@@ -40,8 +42,8 @@ class BinaryTreeLSTM[T: ClassTag](
   withGraph: Boolean = true
 )(implicit ev: TensorNumeric[T])
   extends TreeLSTM[T](inputSize, hiddenSize) {
-  val composer: Module[T] = createComposer()
-  val leafModule: Module[T] = createLeafModule()
+  var composer: Module[T] = createComposer()
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BinaryTreeLSTM.scala]: can we wrap it in a method?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Module.scala [hunk 2]
```diff
@@ -38,10 +39,25 @@ object Module {
    * @tparam T numeric type
    * @return model loaded from path
    */
+  @deprecated
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Module.scala]: please add parameters to the annotation. Or it will throw an warning

**Comment by yiheng** [comment]
As discussed, we should not override each factory method.

**Pull request closed without merging**